File: src/rht/raytracer/Ray.java
package rht.raytracer;

public class Ray {
    private Vec3 origin;
    /** Unit vector. */
    private Vec3 direction;

    public Ray(Vec3 origin, Vec3 direction) {
        this.origin = origin;
        this.direction = direction.normalise();
    }

    public Vec3 getOrigin() {
        return this.origin;
    }

    public Vec3 getDirection() {
        return this.direction;
    }

    public Vec3 distanceAlong(double distance) {
        return this.origin.plus(this.direction.times(distance));
    }
}

File: rustray/src/ray.rs
use crate::maths::{matrix::Matrix, vec::Vector};

#[derive(Clone, Debug, PartialEq)]
pub struct Ray {
    pub origin: Vector<3>,
    /// Unit vector.
    pub direction: Vector<3>,
}

impl Ray {
    pub fn new(origin: Vector<3>, direction: Vector<3>) -> Self {
        Self {
            origin,
            direction: direction.normalise(),
        }
    }

    pub fn distance_along(&self, distance: f64) -> Vector<3> {
        self.origin + (self.direction * distance)
    }

    pub fn transform(&self, transformation: &Matrix<4, 4>) -> Self {
        Ray::new(
            transformation * self.origin,
            transformation.linear_times(self.direction),
        )
    }
}
